### ISSUE TITLE
How to get Google provider ID token and not access token

### ISSUE THREAD

**Issue by vladfulgeanu** [nextauthjs/next-auth#393]
When using the Google provider I want to be able to authenticate with a backend server and for this I need to send the `ID token` to that API and get back the access token. Further info about this flow can be found [here](https://developers.google.com/identity/sign-in/web/backend-auth).

This extra request I plan to do in the JWT callback, but in there I cannot access `ID token`. Only `accessToken` is given back in `token.account`.

This is what I get in the JWT callback:
token (first argument):
```
{
  user: {
    name: <name>,
    email: <email>,
    image: <image>
  },
  account: {
    provider: 'google',
    type: 'oauth',
    id: <id>,
    refreshToken: undefined,
    accessToken: <accessToken>,
    accessTokenExpires: null
  },
  isNewUser: undefined
}
```

profile (second argument):
```
{
  id: <id>,
  email: <email>,
  verified_email: true,
  name: <name>,
  given_name: <given_name>,
  family_name: <family_name>,
  picture: <picture>,
  locale: 'en'
}
```

* [ ] Found the documentation helpful
* [ ] Found documentation but was incomplete
* [ ] Could not find relevant documentation
* [ ] Found the example project helpful
* [X] Did not find the example project helpful

**Comment by iaincollins** [comment]
Yeah Google is unusual and this has come up before, we should really cover this in the docs.

tl;dr Google only returns a `RefreshToken` on first sign in and should to be used with a database. You can force it to issue a new one every time using `access_type=offline` and `prompt=consent` but that is intended for mobile and desktop applications and changes the sort of prompt a user sees when they sign in (and isn't as seamless).

More info in #269

**Comment by vladfulgeanu** [comment]
> access_type

**Comment by vladfulgeanu** [comment]
If I were to use `react-google-login` library, _in the responseGoogle(response) {...} callback function, I should get back a standard JWT located at response.tokenId or res.getAuthResponse().id_token_ (from [here](https://github.com/anthonyjgrove/react-google-login#onsuccess-callback))
That is the only thing that I need in order to send the request to the backend server. I'm not sure how RefreshToken can help me in this case.

**Comment by iaincollins** [comment]
You can try setting `idToken: true` [as a provider option](https://next-auth.js.org/configuration/providers#options) - it works on quite a few providers, but I'm not sure about Google - and then using the [JWT callback](https://next-auth.js.org/configuration/callbacks#jwt) (assuming you are using JSON Web Tokens) to handle saving data from the profile in your JWT to use in subsequent calls.

**Comment by vladfulgeanu** [comment]
I did try exactly this as I saw it possible on the Okta provider, but 2 things happen:
- I get this error:
```
[next-auth][error][OAUTH_CALLBACK_HANDLER_ERROR] [
  Error: Missing or invalid provider account
      at /home/vlad/work/node_modules/next-auth/dist/server/lib/callback-handler.js:32:15
      at Generator.next (<anonymous>)
      at asyncGeneratorStep (/home/vlad/work/node_modules/next-auth/dist/server/lib/callback-handler.js:20:103)
      at _next (/home/vlad/work/node_modules/next-auth/dist/server/lib/callback-handler.js:22:194)
      at /home/vlad/work/node_modules/next-auth/dist/server/lib/callback-handler.js:22:364
      at new Promise (<anonymous>)
      at /home/vlad/work/node_modules/next-auth/dist/server/lib/callback-handler.js:22:97
      at /home/vlad/work/node_modules/next-auth/dist/server/lib/callback-handler.js:185:17
      at /home/vlad/work/node_modules/next-auth/dist/server/routes/callback.js:91:54
      at Generator.next (<anonymous>)
      at asyncGeneratorStep (/home/vlad/work/node_modules/next-auth/dist/server/routes/callback.js:26:103)
      at _next (/home/vlad/work/node_modules/next-auth/dist/server/routes/callback.js:28:194)
      at processTicksAndRejections (internal/process/task_queues.js:93:5)
]
```
The JWT callback is not reached in this case, only the signin callback
- I do get properties of the ID token in the signin callback (iss, sub, iat, exp...), but what I need is the exact ID token string that encodes all these. Is there a method to go from the ID token object, with the iss, sub, iat, exp properties to the ID token string, like base 64 encoding? Best thing would be to be able to access the ID token string directly though.

**Comment by iaincollins** [comment]
Thanks for trying that and reporting what you saw.

I think at the moment the answer is this is not possible, but it seems like a reasonable feature request.

**Comment by Dreched** [comment]
Any update on this? 

The Google provider isn't really complete without it since most front ends will be calling services that need to verify a user's id_token.

**Comment by MaryJJ** [comment]
I need this too, thanks

**Comment by theapplefolks** [comment]
Yeah, any updates? I also really need this.

**Comment by balazsorban44** [comment]
So if I understand correctly, #837 would help here, right?

**Comment by Dreched** [comment]
> So if I understand correctly, #837 would help here, right? Can anyone confirm this? (@vladfulgeanu @Dreched @MaryJJ @theapplefolks)
> 
> It basically sends the `id_token` to the `jwt` callback if it is provided by the IdP at sign-in.

I think #837 will help.  To clarify, `/api/auth/session` will also get `id_token` set from signIn?

**Comment by balazsorban44** [comment]
What ends up in your session is up to you. If you want the `id_token` available in your session, you can look at the `jwt` and `session` callbacks: https://next-auth.js.org/configuration/callbacks

**Comment by fedeberon** [comment]
`let IdToken = async (refreshToken) => {
    var requestOptions = {
        method: 'POST',
        redirect: 'follow'
    };
    
    const response = await fetch(`https://oauth2.googleapis.com/token?refresh_token=${refreshToken}&client_id=[yout-client-id]&client_secret=[your-client-secret]&grant_type=refresh_token&Content-Type=application/x-www-form-urlencoded`, requestOptions);
    let result= await response.json();

    return result.id_token;
}`

**Comment by balazsorban44** [comment]
@fedeberon so when using a refresh token, you will actually receive a new ID token as well? It never occurred to me, I'll check this out, thanks!
